aip/general/0216.md

```diff
@@ -123,6 +123,44 @@ message PublishBookRequest {
 - The comment for the field **should** document the resource pattern.
 - Other fields **may** be included.
 
+### Declarative-friendly state transitions
+
+If a resource's state can be modified by [declarative tooling][aip-128], it
+**should** expose a separate `desired_state` field taking the same enum type.
+
+```proto
+message Book {
+  enum State {
+    STATE_UNSPECIFIED = 0;
+    UNPUBLISHED = 1;
+    PUBLISHING = 2;
+    PUBLISHED = 3;
+  }
+  
+  State state = 1 [(google.api.field_behaviopr = OUTPUT_ONLY)];
+  State desired_state = 2;
+}
+```
+
+- The `state` field's output **may** change to more granular
+  [active states](#common-states) to indicate progress toward the `desired_state`.
+- The `desired_state` field **must** only allow [resting state](#common-states)
+  enum values.
+- Attempting to set `desired_state` to an [active state](#common-states) **must**
+  error with `INVALID_ARGUMENT` (HTTP 400).
+- Attempting to set `desired_state` to a state that cannot be transitioned to
+  from the current state **must** error with `FAILED_PRECONDITION` (HTTP 400).
+- A [long-running][] operation **must not** complete until the `state` has
+  reached the `desired_state` (or **must** error if it cannot).
+- If a resource supports both declarative `desired_state` and imperative
+  [state transition methods](#state-transition-methods), executing a state
+  transition method **should** modify the desired state appropriately.
```

Comment:
This will create drift in declarative clients. It would be better to not have the desired state field in that case.

Reply:
Yes, it will create drift - but I'm trying to consider:

1. Existing APIs with custom methods that might migrate to this pattern and have both for that reason.
2. Flexibility for clients to use a more imperative state transition method if they prefer (presumably the client is not generally declarative in this scenario).

Declarative tooling expects to be the only manager of a given resource, so out of band updates of any kind create drift problems. This is something that API consumers have to manage for themselves.

Reply:
maybe this is why it's best to not have `desired_state`, but rather individual flags that limit state transitions?

e.g. `suspended` would set the resource to be suspended, and not allow other state transitions. 

Or maybe have a way to clear desired state (unspecified) that allows custom methods to mutate the state? effectively a lock.